src/main/java/com/mongodb/kafka/connect/sink/processor/TimeseriesTimeFieldAutoConversion.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.kafka.connect.sink.processor;
+
+import static com.mongodb.kafka.connect.sink.MongoSinkTopicConfig.TIMESERIES_TIMEFIELD_AUTO_CONVERSION_DATE_FORMAT_CONFIG;
+import static com.mongodb.kafka.connect.sink.MongoSinkTopicConfig.TIMESERIES_TIMEFIELD_CONFIG;
+import static java.lang.String.format;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.ResolverStyle;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.bson.BsonDateTime;
+import org.bson.BsonValue;
+
+import com.mongodb.kafka.connect.sink.MongoSinkTopicConfig;
+import com.mongodb.kafka.connect.sink.converter.SinkDocument;
+
+class TimeseriesTimeFieldAutoConversion extends PostProcessor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostProcessor.class);
+
+  private final String fieldName;
+  private final DateTimeFormatter formatter;
+
+  TimeseriesTimeFieldAutoConversion(final MongoSinkTopicConfig config) {
+    super(config);
+    fieldName = config.getString(TIMESERIES_TIMEFIELD_CONFIG);
+    formatter =
+        DateTimeFormatter.ofPattern(
+                config.getString(TIMESERIES_TIMEFIELD_AUTO_CONVERSION_DATE_FORMAT_CONFIG))
+            .withResolverStyle(ResolverStyle.LENIENT);
+  }
+
+  @Override
+  public void process(final SinkDocument doc, final SinkRecord orig) {
+    doc.getValueDoc()
+        .filter(d -> d.containsKey(fieldName) && (d.isNumber(fieldName) || d.isString(fieldName)))
+        .ifPresent(
+            d -> {
+              BsonValue timeField = d.get(fieldName);
+              Optional<BsonDateTime> convertedValue;
+              if (timeField.isNumber()) {
+                convertedValue =
+                    tryToConvert(() -> new BsonDateTime(timeField.asNumber().longValue()));
+              } else {
+                convertedValue =
+                    tryToConvert(
+                        () -> {
+                          Instant dateTime =
+                              LocalDateTime.parse(timeField.asString().getValue(), formatter)
+                                  .toInstant(ZoneOffset.UTC);
```

Comment:
We need to document that configured format is used for parsing strings via `LocalDateTime.parse` because this does have functional implications (or we can document the implications):

1. A parsed string must have not only date information but also time information to be parsed successfully regardless of the format specified.
2. The time offset present in a parsed string is ignored regardless of the format specified.

While the first implication is obvious, the second may not be as obvious, so tested it. Consider the format `uuuu-MM-dd['T'][ ]HH:mm:ss[.][SSSSSS][SSS][X]`, which effectively differs from the default one only by allowing any zone offsets (`[X]`) as opposed to allowing only no or explicit zero offset (`['Z']`):

```java
DateTimeFormatter formatter = DateTimeFormatter
        .ofPattern("uuuu-MM-dd['T'][ ]HH:mm:ss[.][SSSSSS][SSS][X]")
        .withResolverStyle(ResolverStyle.LENIENT); // or STRICT
String text = "1970-01-01T00:00:00.000-0100";
System.out.println(LocalDateTime.parse(text, formatter).toInstant(ZoneOffset.UTC).getEpochSecond());
System.out.println(OffsetDateTime.parse(text, formatter).toInstant().getEpochSecond());
```

This code prints
```
0
3600
```
where 3600 is what a user would expect as a result of parsing, but 0 is what he gets because of us using `LocalDateTime`.

Reply:
Updated.

Reply:
👍